rfc/0000-batch-job-groups.rst

```diff
@@ -0,0 +1,1179 @@
+batches that have been completed (see below). In addition, this design
+means that a cancellation of one job group has temporarily prevented
+the entire batch from being seen by the autoscaler and scheduler. I
+think for the most common use case, this constraint is okay. Most
+batches are small and the QoB use case has all running jobs in the
+same job group so there is no change in behavior from what we
+currently do.
+
+
+~~~~~~~~~~~~~~
+Job Completion
+~~~~~~~~~~~~~~
+
+When a job is marked complete, all job groups the job is a member of
+are checked to see if the number of jobs in the job group is equal to
+the number completed. We are guaranteed that the job that sees the
+number of jobs equals the number completed is the last job to complete
+despite no locking being done. We then execute the callback for any
+newly completed job groups. The amount of extra overhead in the mark
+job complete SQL procedure compared to what we have now is
+O(n_job_groups) the job is a member of, which will need some sort of
+bound on it. This is because we have to update values in the billing
+tables and the table that keeps track of the job states per job group
+for each job group the job is a member of. When the batch is
+completed, we will delete the extra rows from the
+``user_inst_coll_resources`` to make sure that table is as fast as
+possible (O(n_active_batches)).
+
+
+~~~~~~~~~~~~~~~
+Job Group Trees
+~~~~~~~~~~~~~~~
+
+A job group tree consists of a tree structure where a job group can
+have children job groups where each child job group has a partition of
+the jobs in the parent job group. Therefore, a job is a member of its
+specific job group plus all of the parent job groups forming a tree
+structure. Each job group is identified by a path that starts with "/"
+which represents the root job group. The implementation consists of
+two tables that are used to perform operations on the tree and map a
+path identifier to the job group of interest:
+
+- ``job_group_tree``
+- ``job_group_tree_parents``
+
+The implementation of cancellation for job groups in the job group
+tree is to also cancel any children job groups by simply inserting the
+child job group IDs into the ``job_groups_cancelled``
+table. Aggregations for billing and job states propagating up the job
+group tree are taken care of automatically as we've densely defined a
+job group being a member of all job groups including the parents.
+
+This additional layer can be implemented **later** on as it is not
+crucial for QoB functionality. Instead, it will provide a nicer user
+experience for both QoB and regular Hail Batch users.
+
+
+--------
+Examples
+--------
+
+Although QoB is the primary use case for this feature, we will use the
+Python client interface implemented in ``aioclient.py`` in order to
+demonstrate the utility of this feature. The examples below are for
+the longer term vision. We do not have to implement all of this
+functionality right away.
```

Comment:
Have an examples section which only covers this RFC and then we can put another examples section under future work which considers the full scope.